src/text/raw_text_reader.rs

```diff
@@ -1044,4 +1057,58 @@ mod reader_tests {
 
         Ok(())
     }
+
+    #[test]
+    fn structs_trailing_comma() -> IonResult<()> {
+        let pretty_ion = br#"
+            // Structs with last field with/without trailing comma
+            (
+                {a:1, b:2,}     // with trailing comma
+                {a:1, b:2 }     // without trailing comma
+            )
+        "#;
+        let mut reader = RawTextReader::new(&pretty_ion[..]);
+        assert_eq!(reader.next()?, RawStreamItem::Value(IonType::SExpression));
+        reader.step_in()?;
+        assert_eq!(reader.next()?, RawStreamItem::Value(IonType::Struct));
+
+        reader.step_in()?;
+        assert_eq!(reader.next()?, RawStreamItem::Value(IonType::Integer));
+        assert_eq!(reader.field_name()?, RawSymbolToken::Text("a".to_string()));
+        assert_eq!(reader.read_i64()?, 1);
+        assert_eq!(reader.next()?, RawStreamItem::Value(IonType::Integer));
+        assert_eq!(reader.field_name()?, RawSymbolToken::Text("b".to_string()));
+        assert_eq!(reader.read_i64()?, 2);
+        reader.step_out()?;
+
+        assert_eq!(reader.next()?, RawStreamItem::Value(IonType::Struct));
+        reader.step_out()?;
+        Ok(())
+    }
+
+    #[test]
+    fn annotation_false() -> IonResult<()> {
+        // The reader will reject the unquoted boolean keyword 'false' when used as an annotation
+        let pretty_ion = br#"
+            false::23
+        "#;
+        let mut reader = RawTextReader::new(&pretty_ion[..]);
+        let result = reader.next();
+        println!("{:?}", result);
+        assert!(result.is_err());
+        Ok(())
+    }
+
+    #[test]
+    fn annotation_nan() -> IonResult<()> {
+        // The reader will reject the unquoted float keyword 'nan' when used as an annotation
+        let pretty_ion = br#"
+            nan::23
+        "#;
+        let mut reader = RawTextReader::new(&pretty_ion[..]);
+        let result = reader.next();
+        println!("{:?}", result);
+        assert!(result.is_err());
+        Ok(())
+    }
```

Comment:
I found these two mildly surprising, so I went to go read first the [Type Annotations](https://amzn.github.io/ion-docs/docs/spec.html#annot) section of the Ion specification and then the [text ANTLR grammar](https://amzn.github.io/ion-docs/grammar/IonText.g4.txt).  The spec didn't call out this kind of limitation, and my read of the ANTLR grammar (disclaimer I could be bad at this, relevant fragment reproduced below) doesn't seem to support this either. 


```plain
top_level_value
    : annotation+ top_level_value
    | delimiting_entity
    // numeric literals (if followed by something), need to be followed by
    // whitespace or a token that is either quoted (e.g. string) or
    // starts with punctuation (e.g. clob, struct, list)
    | numeric_entity ws
    | numeric_entity quoted_annotation value
    | numeric_entity delimiting_entity
    // literals that are unquoted symbols or keywords have a similar requirement
    // as the numerics above, they have different productions because the
    // rules for numerics are the same in s-expressions, but keywords
    // have different rules between top-level and s-expressions.
    | keyword_entity ws
    | keyword_entity quoted_annotation value
    | keyword_entity keyword_delimiting_entity
    ;

annotation
    : symbol ws* COLON COLON ws*
    ;

symbol
    : IDENTIFIER_SYMBOL
    // note that this is because we recognize the type names for null
    // they are ordinary symbols on their own
    | TYPE
    | QUOTED_SYMBOL
    ;

IDENTIFIER_SYMBOL
    : [$_a-zA-Z] ([$_a-zA-Z] | DEC_DIGIT)*
    ;

numeric_entity
    : BIN_INTEGER
    | DEC_INTEGER
    | HEX_INTEGER
    | TIMESTAMP
    | FLOAT
    | DECIMAL
    ;
```

*Is* this how annotations *should* work?

Reply:
In cases like this, I tend to defer to `ion-tests`. Here are some relevant tests from [the `bad` examples directory](https://github.com/amzn/ion-tests/tree/master/iontestdata/bad):
* [`false` as an annotation](https://github.com/amzn/ion-tests/blob/master/iontestdata/bad/annotationFalse.ion)
* [`nan` as an annotation](https://github.com/amzn/ion-tests/blob/master/iontestdata/bad/annotationNan.ion)

`ion-java` does not allow unquoted keywords as annotations, which jibes with these `ion-tests`.